sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/Changelog.scala

```diff
@@ -0,0 +1,153 @@
+  extends Iterator[ByteArrayPair] with Logging {
+  private def decompressStream(inputStream: DataInputStream): DataInputStream = {
+    val compressed = compressionCodec.compressedInputStream(inputStream)
+    new DataInputStream(compressed)
+  }
+  private val sourceStream = try {
+    fm.open(fileToRead)
+  } catch {
+    case f: FileNotFoundException =>
+      throw new IllegalStateException(
+        s"Error reading streaming state file of $this: $fileToRead does not exist. " +
+          s"If the stream job is restarted with a new or updated state operation, please" +
+          s" create a new checkpoint location or clear the existing checkpoint location.", f)
+  }
+  private val input: DataInputStream = decompressStream(sourceStream)
+  private var byteArrayPair: ByteArrayPair = null
+  private var eof = false
+
+  override def hasNext: Boolean = {
+    maybeReadNext()
+    byteArrayPair != null
+  }
+
+  override def next(): ByteArrayPair = {
+    maybeReadNext()
+    val nextByteArrayPair = byteArrayPair
+    byteArrayPair = null
+    nextByteArrayPair
+  }
+
+  def close(): Unit = { if (input != null) input.close() }
+
+  private def maybeReadNext(): Unit = {
+    if (!eof && byteArrayPair == null) {
+      val keySize = input.readInt()
+      if (keySize == -1) {
+        eof = true
+      } else if (keySize < 0) {
+        throw new IOException(
+          s"Error reading streaming state file $fileToRead of $this: key size cannot be $keySize")
+      } else {
+        val keyBuffer = new Array[Byte](keySize)
+        ByteStreams.readFully(input, keyBuffer, 0, keySize)
+        val valueSize = input.readInt()
+        if (valueSize < 0) {
+          byteArrayPair = new ByteArrayPair(keyBuffer, null)
+        } else {
+          val valueBuffer = new Array[Byte](valueSize)
+          ByteStreams.readFully(input, valueBuffer, 0, valueSize)
+          byteArrayPair = new ByteArrayPair(keyBuffer, valueBuffer)
+        }
+      }
+    }
+  }
+}
+
+class ChangelogWriter(fm: CheckpointFileManager, file: Path,
+                      compressionCodec: CompressionCodec) extends Logging {
+  private def compressStream(outputStream: DataOutputStream): DataOutputStream = {
+    val compressed = compressionCodec.compressedOutputStream(outputStream)
+    new DataOutputStream(compressed)
+  }
+
+  private var backingFileStream: CancellableFSDataOutputStream =
+    fm.createAtomic(file, overwriteIfPossible = true)
+  private var compressedStream: DataOutputStream = compressStream(backingFileStream)
+
+  def put(key: Array[Byte], value: Array[Byte]): Unit = {
+    assert(compressedStream != null)
+    compressedStream.writeInt(key.size)
+    compressedStream.write(key)
+    compressedStream.writeInt(value.size)
+    compressedStream.write(value)
+  }
+
+  def delete(key: Array[Byte]): Unit = {
+    assert(compressedStream != null)
+    compressedStream.writeInt(key.size)
+    compressedStream.write(key)
+    compressedStream.writeInt(-1)
```

Comment:
Curious, why are we adding this ?

Reply:
A -1 key length signals the end of files, which is the same format as hdfs state store changelog.

Reply:
Right - but for delete, its trying to signal like empty value ?

Reply:
No, for delete, we have | key length | key bytes | value length = -1 |, 
key length is non-negative for both put and delete, a -1 in the key length field is end of file.